portfolio/src/main/java/com/google/sps/servlets/SpotifyServlet.java

```diff
@@ -0,0 +1,150 @@
+    .setClientId(clientId)
+    .setClientSecret(clientSecret)
+    .setRedirectUri(redirectUri)
+    .build();
+  
+  private static final AuthorizationCodeUriRequest authorizationCodeUriRequest = spotifyApi.authorizationCodeUri()
+    .scope("user-read-currently-playing")
+    .show_dialog(true)
+    .build();
+ 
+  private static AuthorizationCodeRequest authorizationCodeRequest = spotifyApi.authorizationCode(oauthToken)
+    .build();
+
+  private static GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
+    .getUsersCurrentlyPlayingTrack()
+    .build();
+
+  private static AuthorizationCodeRefreshRequest authorizationCodeRefreshRequest = spotifyApi.authorizationCodeRefresh()
+    .build();
+
+  public static String authorizationCodeUri() {
+    final URI uri = authorizationCodeUriRequest.execute();
+    return uri.toString();
+  }
+
+  public static String formatAsMinutesAndSeconds(Integer milliseconds) {
+    Integer seconds = (Integer)(milliseconds/ 1000) % 60;
+    Integer minutes = (Integer)(milliseconds/ (1000) / 60);
+    return String.format("%d:%02d", minutes, seconds);
+  }
+
+  public static void authorizationCode() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRequest.execute();
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static void authorizationCodeRefresh() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRefreshRequest.execute();
+ 
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+ 
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static String getCurrentTrack() {
+    try {
+      final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();
+
+      final CurrentlyPlaying currentlyPlaying = currentlyPlayingFuture.join();
+
+      if (currentlyPlaying == null) {
+        return "No song currently playing.";
+      }
+
+      IPlaylistItem item = currentlyPlaying.getItem();
+      String songName = "";
+      String artists = "";
+      
+      if (item instanceof Track) {
+        songName = ((Track)item).getName();
+
+        for (int iterator = 0; iterator < ((Track)item).getArtists().length; iterator++) {
```

Comment:
nit: This is the one case where a 1-letter variable name is allowed and actually preferred :) Also 'i' stands for 'index' not 'iterator', those are different

Reply:
Thank you for pointing that out! I'll make sure to fix that and not do it again.